### ISSUE TITLE
Support Supplier<Arbitrary> as Parameter in `@ForAll` and `@From` Annotations

### ISSUE THREAD

**Issue by jlink** [jlink/jqwik#91]
> Such a parameter for `@ForAll` would enable type-safe references to Arbitrary-Providers as well as IDE navigation support.

@mmerdes Moved over from other repository

see https://github.com/jlink/pbt-workshop/issues/1

**Comment by jlink** [comment]
Thanks for the suggestion. 

I assume you habe something like that in mind:

```java
@Property
void withSupplier(@ForAll(supplier = GermanZipCodes.class) String zipCode) {
	assertThat(zipCode).hasSize(5);
}

class GermanZipCodes implements Supplier<Arbitrary<?>> {
	@Override
	public Arbitrary<String> get() {
		return Arbitraries.strings().withCharRange('0', '9').ofLength(5);
	}
}
```

However, to give a supplier more power and to get it closer to what an `ArbitraryProvider` could do, I see a more elaborate version:

```java
@Property
void withSupplier(
	@ForAll(supplier = Money.class) BigDecimal moneyAsBigDecimal,
	@ForAll(supplier = Money.class) double moneyAsDouble
) {
	assertThat(moneyAsBigDecimal).isBetween(BigDecimal.ZERO, BigDecimal.valueOf(100));
	assertThat(moneyAsDouble).isBetween(0.0, 100.0);
}

class Money implements Function<TypeUsage, Arbitrary<?>> {
	@Override
	public Arbitrary<?> apply(TypeUsage targetType) {
		if (targetType.isAssignableFrom(BigDecimal.class))
			return Arbitraries.bigDecimals().between(BigDecimal.ZERO, BigDecimal.valueOf(100));
		if (targetType.isAssignableFrom(Double.class))
			return Arbitraries.doubles().between(0.0, 100.0);
		return null;
	}
}
```

And even the full move to accepting `ArbitraryProvider` would be doable:

```java
@Property
void withSupplier(
	@ForAll(supplier = Money.class) BigDecimal moneyAsBigDecimal,
	@ForAll(supplier = Money.class) double moneyAsDouble
) {
	assertThat(moneyAsBigDecimal).isBetween(BigDecimal.ZERO, BigDecimal.valueOf(100));
	assertThat(moneyAsDouble).isBetween(0.0, 100.0);
}

class Money implements ArbitraryProvider {
	@Override
	public boolean canProvideFor(TypeUsage targetType) {
		return true;
	}

	@Override
	public Set<Arbitrary<?>> provideFor(
		TypeUsage targetType,
		SubtypeProvider subtypeProvider
	) {
		if (targetType.isAssignableFrom(BigDecimal.class))
			return Collections.singleton(...);
		if (targetType.isAssignableFrom(Double.class))
			return Collections.singleton(...);
		return Collections.emptySet();
	}
}
```
That would give the full capability of returning more than one arbitrary and to handle parameterized types. With a subinterface like that:

```java
interface ArbitrarySupplier extends ArbitraryProvider, Supplier<Arbitrary<?>> {
	@Override
	default boolean canProvideFor(TypeUsage targetType) {
		return true;
	}

	@Override
	default Set<Arbitrary<?>> provideFor(TypeUsage targetType, SubtypeProvider subtypeProvider) {
		return Collections.singleton(this.get());
	}
}
```

the simplicity of the first example could also be maintained.

What do you think?

BTW, the same functionality should then also be usable through `@From(supplier=...)` which allows to provide generators for type parameters of a generic type.

**Comment by slaymaker1907** [comment]
Have you considered creating a Domain for this? Domains are a really powerful way to do similar things to this.

**Comment by jlink** [comment]
Could you elaborate with an example?

**Comment by jlink** [comment]
I tend to turn the dependencies around:

```
interface ArbitrarySupplier extends Supplier<Arbitrary<?>> {
    default Set<Arbitrary<?>> supply(
        TypeUsage targetType,
	SubtypeProvider subtypeProvider
    ) {
        return Collections.singleton(get());
    }

   default Arbitrary<?> get() {
       throw new JqwikException("You have to override either ArbitrarySupplier.get() or ArbitrarySupplier.supply()");
   }
}

interface ArbitraryProvider extends ArbitrarySupplier {
   default Set<Arbitrary<?>> supply(...) { return provideFor(...); }
}
```

**Comment by jlink** [comment]
Available in "1.6.3-SNAPSHOT".

**Comment by jlink** [comment]
User guide documentation: https://jqwik.net/docs/snapshot/user-guide.html#arbitrary-suppliers
